Multi config dependent components (VSC-1480)

### OS

Windows

### Operating System version

Windows 10 Pro

### Visual Studio Code version

1.92.2

### ESP-IDF version

5.1.4

### Python version

3.11.7

### Doctor command output

---------------------------------------------- ESP-IDF Extension for Visual Studio Code report ---------------------------------------------
OS win32 x64 10.0.19045 
System environment variable IDF_PYTHON_ENV_PATH 
 undefined 
System environment variable PATH 
 C:\Program Files (x86)\Common Files\Oracle\Java\javapath;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Git\cmd;C:\Program Files\PuTTY\;C:\Program Files\IVI Foundation\VISA\Win64\bin;C:\Program Files (x86)\IVI Foundation\VISA\WinNT\bin;C:\Program Files (x86)\IVI Foundation\IVI\bin;C:\Program Files\IVI Foundation\IVI\bin;C:\Program Files (x86)\IVI Foundation\VISA\WinNT\Bin\;C:\Program Files\IVI Foundation\VISA\Win64\Bin\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\msys64\mingw64\bin;C:<HOMEPATH>\AppData\Local\Microsoft\WindowsApps;C:<HOMEPATH>\AppData\Local\Programs\Microsoft VS Code\bin;C:\CMake\bin;D:\Downloads\make-3.81-bin\bin; 
System environment variable PYTHON 
 undefined 
Visual Studio Code version 1.92.2 
Visual Studio Code language en 
Visual Studio Code shell C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe 
ESP-IDF Extension version 1.8.0 
Workspace folder -=MY_PATH=-\IDF-MultiConfig\multi_config 
---------------------------------------------------- Extension configuration settings ------------------------------------------------------
ESP-ADF Path (idf.espAdfPath) ${env:ADF_PATH}
ESP-IDF Path (idf.espIdfPath) c:<HOMEPATH>\esp\v5.1\esp-idf
ESP-MDF Path (idf.espMdfPath) ${env:MDF_PATH}
ESP-Matter Path (idf.espMatterPath) ${env:ESP_MATTER_PATH}
ESP-HomeKit-SDK Path (idf.espHomeKitSdkPath) ${env:HOMEKIT_PATH}
Custom extra paths (idf.customExtraPaths) c:\esp-idf5.1\tools\xtensa-esp-elf-gdb\12.1_20231023\xtensa-esp-elf-gdb\bin;c:\esp-idf5.1\tools\riscv32-esp-elf-gdb\12.1_20231023\riscv32-esp-elf-gdb\bin;c:\esp-idf5.1\tools\xtensa-esp32-elf\esp-12.2.0_20230208\xtensa-esp32-elf\bin;c:\esp-idf5.1\tools\xtensa-esp32s2-elf\esp-12.2.0_20230208\xtensa-esp32s2-elf\bin;c:\esp-idf5.1\tools\xtensa-esp32s3-elf\esp-12.2.0_20230208\xtensa-esp32s3-elf\bin;c:\esp-idf5.1\tools\riscv32-esp-elf\esp-12.2.0_20230208\riscv32-esp-elf\bin;c:\esp-idf5.1\tools\esp32ulp-elf\2.35_20220830\esp32ulp-elf\bin;c:\esp-idf5.1\tools\cmake\3.24.0\bin;c:\esp-idf5.1\tools\openocd-esp32\v0.12.0-esp32-20240318\openocd-esp32\bin;c:\esp-idf5.1\tools\ninja\1.10.2;c:\esp-idf5.1\tools\idf-exe\1.0.3;c:\esp-idf5.1\tools\ccache\4.8\ccache-4.8-windows-x86_64;c:\esp-idf5.1\tools\dfu-util\0.11\dfu-util-0.11-win64;c:\esp-idf5.1\tools\esp-rom-elfs\20230320
Custom extra vars (idf.customExtraVars)
    OPENOCD_SCRIPTS: c:\esp-idf5.1\tools\openocd-esp32\v0.12.0-esp32-20240318/openocd-esp32/share/openocd/scripts
    IDF_CCACHE_ENABLE: 1
    ESP_ROM_ELF_DIR: c:\esp-idf5.1\tools\esp-rom-elfs\20230320/
Virtual env Python Path (idf.pythonBinPath) c:\esp-idf5.1\python_env\idf5.1_py3.11_env\Scripts\python.exe
Serial port (idf.port) COM1
OpenOCD Configs (idf.openOcdConfigs) board/esp32s3-builtin.cfg
ESP-IDF Tools Path (idf.toolsPath) c:\esp-idf5.1
Git Path (idf.gitPath) c:\esp-idf5.1\tools\idf-git\2.39.2\cmd\git.exe
Notification Mode (idf.notificationMode) All
-------------------------------------------------------- Configurations access -------------------------------------------------------------
Access to ESP-ADF Path (idf.espAdfPath) false
Access to ESP-IDF Path (idf.espIdfPath) true
Access to ESP-MDF Path (idf.espMdfPath) false
Access to ESP-Matter Path (idf.espMatterPath) false
Access to ESP-HomeKit Path (idf.espHomeKitSdkPath) false
Access to ESP-IDF Custom extra paths
Access to c:\esp-idf5.1\tools\xtensa-esp-elf-gdb\12.1_20231023\xtensa-esp-elf-gdb\bin: true
Access to c:\esp-idf5.1\tools\riscv32-esp-elf-gdb\12.1_20231023\riscv32-esp-elf-gdb\bin: true
Access to c:\esp-idf5.1\tools\xtensa-esp32-elf\esp-12.2.0_20230208\xtensa-esp32-elf\bin: true
Access to c:\esp-idf5.1\tools\xtensa-esp32s2-elf\esp-12.2.0_20230208\xtensa-esp32s2-elf\bin: true
Access to c:\esp-idf5.1\tools\xtensa-esp32s3-elf\esp-12.2.0_20230208\xtensa-esp32s3-elf\bin: true
Access to c:\esp-idf5.1\tools\riscv32-esp-elf\esp-12.2.0_20230208\riscv32-esp-elf\bin: true
Access to c:\esp-idf5.1\tools\esp32ulp-elf\2.35_20220830\esp32ulp-elf\bin: true
Access to c:\esp-idf5.1\tools\cmake\3.24.0\bin: true
Access to c:\esp-idf5.1\tools\openocd-esp32\v0.12.0-esp32-20240318\openocd-esp32\bin: true
Access to c:\esp-idf5.1\tools\ninja\1.10.2: true
Access to c:\esp-idf5.1\tools\idf-exe\1.0.3: true
Access to c:\esp-idf5.1\tools\ccache\4.8\ccache-4.8-windows-x86_64: true
Access to c:\esp-idf5.1\tools\dfu-util\0.11\dfu-util-0.11-win64: true
Access to c:\esp-idf5.1\tools\esp-rom-elfs\20230320: true
Access to Virtual env Python Path (idf.pythonBinPath) true
Access to CMake in environment PATH undefined
Access to Ninja in environment PATH undefined
Access to ESP-IDF Tools Path (idf.toolsPath) true
-------------------------------------------------------- Configurations has spaces -------------------------------------------------------------
Spaces in system environment Path true
Spaces in ESP-ADF Path (idf.espAdfPath) false
Spaces in ESP-IDF Path (idf.espIdfPath) false
Spaces in ESP-MDF Path (idf.espMdfPath) false
Spaces in ESP-Matter Path (idf.espMatterPath) false
Spaces in ESP-HomeKit-SDK Path (idf.espHomeKitSdkPath) false
Spaces in ESP-IDF Custom extra paths
Spaces in c:\esp-idf5.1\tools\xtensa-esp-elf-gdb\12.1_20231023\xtensa-esp-elf-gdb\bin: false
Spaces in c:\esp-idf5.1\tools\riscv32-esp-elf-gdb\12.1_20231023\riscv32-esp-elf-gdb\bin: false
Spaces in c:\esp-idf5.1\tools\xtensa-esp32-elf\esp-12.2.0_20230208\xtensa-esp32-elf\bin: false
Spaces in c:\esp-idf5.1\tools\xtensa-esp32s2-elf\esp-12.2.0_20230208\xtensa-esp32s2-elf\bin: false
Spaces in c:\esp-idf5.1\tools\xtensa-esp32s3-elf\esp-12.2.0_20230208\xtensa-esp32s3-elf\bin: false
Spaces in c:\esp-idf5.1\tools\riscv32-esp-elf\esp-12.2.0_20230208\riscv32-esp-elf\bin: false
Spaces in c:\esp-idf5.1\tools\esp32ulp-elf\2.35_20220830\esp32ulp-elf\bin: false
Spaces in c:\esp-idf5.1\tools\cmake\3.24.0\bin: false
Spaces in c:\esp-idf5.1\tools\openocd-esp32\v0.12.0-esp32-20240318\openocd-esp32\bin: false
Spaces in c:\esp-idf5.1\tools\ninja\1.10.2: false
Spaces in c:\esp-idf5.1\tools\idf-exe\1.0.3: false
Spaces in c:\esp-idf5.1\tools\ccache\4.8\ccache-4.8-windows-x86_64: false
Spaces in c:\esp-idf5.1\tools\dfu-util\0.11\dfu-util-0.11-win64: false
Spaces in c:\esp-idf5.1\tools\esp-rom-elfs\20230320: false
Spaces in Virtual env Python Path (idf.pythonBinPath) false
Spaces in ESP-IDF Tools Path (idf.toolsPath) false
----------------------------------------------------------- Executables Versions -----------------------------------------------------------
Git version 2.39.2.windows.1
ESP-IDF version 5.1.4
Python version 3.11.2
Python's pip version 24.2
-------------------------------------------------- Project configuration settings ----------------------------------------------------------
-------------------------------------------------- Python packages in idf.pythonBinPath ----------------------------------------------------
annotated-types version: 0.7.0
bitarray version: 2.9.2
bitstring version: 4.2.3
CacheControl version: 0.14.0
certifi version: 2024.6.2
cffi version: 1.16.0
charset-normalizer version: 3.3.2
click version: 8.0.4
colorama version: 0.4.6
construct version: 2.10.70
contextlib2 version: 21.6.0
cryptography version: 39.0.2
ecdsa version: 0.19.0
esp-coredump version: 1.11.0
esp-debug-backend version: 1.0.3
esp-idf-kconfig version: 1.4.2
esp-idf-monitor version: 1.4.0
esp-idf-panic-decoder version: 1.1.0
esp-idf-size version: 1.5.0
esptool version: 4.7.0
filelock version: 3.15.4
freertos-gdb version: 1.0.3
idf-component-manager version: 1.5.3
idna version: 3.7
intelhex version: 2.3.0
jsonref version: 1.1.0
kconfiglib version: 14.1.0
markdown-it-py version: 3.0.0
mdurl version: 0.1.2
msgpack version: 1.0.8
packaging version: 24.1
pip version: 24.2
pycparser version: 2.22
pydantic version: 2.8.2
pydantic_core version: 2.20.1
pydantic-settings version: 2.4.0
pyelftools version: 0.29
pygdbmi version: 0.9.0.2
Pygments version: 2.18.0
pyparsing version: 3.0.9
pyserial version: 3.5
python-dotenv version: 1.0.1
pywin32 version: 306
PyYAML version: 6.0.1
reedsolo version: 1.7.0
requests version: 2.32.3
requests-file version: 1.5.1
requests-toolbelt version: 1.0.0
rich version: 13.7.1
schema version: 0.7.5
setuptools version: 73.0.1
six version: 1.16.0
tqdm version: 4.66.4
typing_extensions version: 4.12.2
urllib3 version: 1.26.19
windows-curses version: 2.3.3
---------------------------------------------------- Check ESP-IDF python requirements.txt -------------------------------------------------
Check ESP-IDF Python packages Python requirements are satisfied.
---------------------------------------------------- Check ESP-IDF debug adapter requirements.txt ------------------------------------------
Check Debug AdapterPython packages Python requirements are satisfied.
----------------------------------------------------------- Latest error -----------------------------------------------------------------
{
    "errno": -4058,
    "code": "ENOENT",
    "syscall": "open",
    "path": "-=MY_PATH=-\\IDF-MultiConfig\\multi_config\\.vscode\\launch.json"
}
--------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------- Logfile -----------------------------------------------------------------

--------------------------------------------------------------------------------------------------------------------------------------------


### Extension

_No response_

### Description

During multi config project development I got an entention to have config dependent components. Dependent in the most possible way - do not compile components not required in current config (not just to save some compilation time, but more for have them more clear of extra #ifdef conditions for all not-applicable parts of project in current config).

I agree that project structure is not very clear and aligned, and in some cases it would be no problem to compile some extra components not participate in linking stage.

But my question is there any other way than I make it, that will make compile only required components possible?

### Debug Message

```plain
IDF-MultiConfig/multi_config/main/multi_config_example_main.c:14:14: fatal error: Component_B.h: No such file or directory
   14 |     #include "Component_B.h"
      |              ^~~~~~~~~~~~~~~
```


### Other Steps to Reproduce

1. As usual, started with [MultiConfig example](https://github.com/espressif/vscode-esp-idf-extension/blob/master/docs/tutorial/multiple_config.md).
2. Add Component_A and Component_B via ESP-IDF menu.
3. I dont want all found componetns to compile, so need to add minimal list in root CMakeLists.txt
```
set(COMPONENTS main bootloader esptool_py)
```
4. Now we need to add components dependency from selected configuration. Only way that works fine for me is
```
if(CONFIG_EXAMPLE_FUNC_IMPL_DEV)

    target_sources(${COMPONENT_LIB} PRIVATE "func_dev.c")
    target_link_libraries(${COMPONENT_LIB} PRIVATE Component_A)
    
elseif(CONFIG_EXAMPLE_FUNC_IMPL_PROD)

    target_sources(${COMPONENT_LIB} PRIVATE "func_prod.c")
    target_link_libraries(${COMPONENT_LIB} PRIVATE Component_B)

endif()
```
We can not use REQUIRES field of idf_component_register() becouse IDF Build system not accepts variables there, and, now it becomes clear, that there is a reson for it - on clear build we need to make sdkconfig file (one will contain CONFIG_EXAMPLE_FUNC_IMPL_PROD definition), that depends on components paticipates in project, and I want to make self-dependent link of dependencies: required components depends on sdkconfig definitions, sdkconfig defenitions depends on used components list. Lock.
UPD: Here is explicit statement about the case in [documentation](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/build-system.html#optional-project-variables:~:text=The%20values%20of%20REQUIRES%20and%20PRIV_REQUIRES%20should%20not%20depend%20on%20any%20configuration%20choices%20(CONFIG_xxx%20macros).%20This%20is%20because%20requirements%20are%20expanded%20before%20the%20configuration%20is%20loaded.%20Other%20component%20variables%20(like%20include%20paths%20or%20source%20files)%20can%20depend%20on%20configuration%20choices.)
6.  Now I got an error No such file or directory #include "Component_B.h" becouse Component_B is not in constant REQUIRES list of any components and it was not added to entire scope due to set(COMPONENTS main bootloader esptool_py) limited scope. target_link_libraries(${COMPONENT_LIB} PRIVATE Component_B) not helps, becouse IDF Build system scans required components onlu on first CMake run, that quits on idf_component_register() found.
7.  So, only way I have to go on - is to keep all my components compilable in all possible configurations and leave all the important bricks picking for linker. Am I right to proceed only that way?





### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

--- a/.github/workflows/pr-comment.yml
+++ b/.github/workflows/pr-comment.yml
@@ -8,7 +8,7 @@ jobs:
     if: github.event.workflow_run.event == 'pull_request' && github.event.workflow_run.conclusion == 'success'
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/github-script@v6
+      - uses: actions/github-script@v7
         with:
           # This snippet is public-domain, taken from
           # https://github.com/oprypin/nightly.link/blob/master/.github/workflows/pr-comment.yml

--- a/docs_espressif/en/additionalfeatures/project-configuration.rst
+++ b/docs_espressif/en/additionalfeatures/project-configuration.rst
@@ -22,8 +22,6 @@ To allow you to have multiple configurations for the same project, you can defin
 +-----------------------------------+-------------------------------------------------------------------------------------------+
 | **idf.customExtraVars**           | Variables to be added to system environment variables                                     |
 +-----------------------------------+-------------------------------------------------------------------------------------------+
-| **idf.adapterTargetName**         | ESP-IDF Target Chip (Example: esp32)                                                      |
-+-----------------------------------+-------------------------------------------------------------------------------------------+
 | **idf.flashBaudRate**             | Flash Baud rate                                                                           |
 +-----------------------------------+-------------------------------------------------------------------------------------------+
 | **idf.monitorBaudRate**           | Monitor Baud Rate (Empty by default to use SDKConfig CONFIG_ESP_CONSOLE_UART_BAUDRATE)    |

--- a/docs_espressif/en/settings.rst
+++ b/docs_espressif/en/settings.rst
@@ -110,7 +110,7 @@ This is how the extension uses them:
 6. **idf.openOcdLaunchArgs**: Launch arguments string array for OpenOCD. The resulting OpenOCD launch command looks like this: ``openocd -d${idf.openOcdDebugLevel} -f ${idf.openOcdConfigs} ${idf.openOcdLaunchArgs}``.
 
 .. note::
-  * When you use the command **ESP-IDF: Set Espressif Device Target** it will override **idf.adapterTargetName** with selected chip and **idf.openOcdConfigs** with its default OpenOCD Configuration Files.
+  * When you use the command **ESP-IDF: Set Espressif Device Target** it will override the current sdkconfig IDF_TARGET with selected Espressif chip and **idf.openOcdConfigs** with its default OpenOCD Configuration Files.
   * If you want to customize the **idf.openOcdConfigs** alone, you can use the **ESP-IDF: Select OpenOCD Board Configuration** or modify your settings.json directly.
 
 Code Coverage Specific Settings

--- a/src/build/buildTask.ts
+++ b/src/build/buildTask.ts
@@ -168,7 +168,7 @@ export class BuildTask {
         }
       }
 
-      const sdkconfigFile = getSDKConfigFilePath(this.currentWorkspace);
+      const sdkconfigFile = await getSDKConfigFilePath(this.currentWorkspace);
       if (compilerArgs.indexOf("SDKCONFIG") === -1) {
         compilerArgs.push(`-DSDKCONFIG=${sdkconfigFile}`);
       }

--- a/src/build/buildTask.ts
+++ b/src/build/buildTask.ts
@@ -23,9 +23,10 @@ import * as vscode from "vscode";
 import * as idfConf from "../idfConfiguration";
 import {
   appendIdfAndToolsToPath,
-  isBinInPath,
-  getEspIdfFromCMake,
   compareVersion,
+  getEspIdfFromCMake,
+  getSDKConfigFilePath,
+  isBinInPath,
 } from "../utils";
 import { TaskManager } from "../taskManager";
 import { selectedDFUAdapterId } from "../flash/dfu";
@@ -165,6 +166,11 @@ export class BuildTask {
         } else {
           compilerArgs.push("-S", this.currentWorkspace.fsPath);
         }
+      }
+
+      const sdkconfigFile = getSDKConfigFilePath(this.currentWorkspace);
+      if (compilerArgs.indexOf("SDKCONFIG") === -1) {
+        compilerArgs.push(`-DSDKCONFIG=${sdkconfigFile}`);
       }
 
       const sdkconfigDefaults =

--- a/src/config.ts
+++ b/src/config.ts
@@ -29,7 +29,6 @@ export namespace ESP {
 
   export namespace ProjectConfiguration {
     export let store: ProjectConfigStore;
-    export const CONFIGURATION_LIST_KEY = "PROJECT_CONFIGURATION_KEYS";
     export const SELECTED_CONFIG = "SELECTED_PROJECT_CONFIG";
     export const PROJECT_CONFIGURATION_FILENAME =
       "esp_idf_project_configuration.json";

--- a/src/espIdf/menuconfig/confServerProcess.ts
+++ b/src/espIdf/menuconfig/confServerProcess.ts
@@ -74,13 +74,14 @@ export class ConfserverProcess {
       const pythonBinPath = await getVirtualEnvPythonPath(workspaceFolder);
       const modifiedEnv = await appendIdfAndToolsToPath(workspaceFolder);
       if (!ConfserverProcess.instance) {
+        const configFile = await getSDKConfigFilePath(workspaceFolder);
         ConfserverProcess.instance = new ConfserverProcess(
           workspaceFolder,
           extensionPath,
           pythonBinPath,
+          configFile,
           modifiedEnv
         );
-        ConfserverProcess.instance.configFile = await getSDKConfigFilePath(workspaceFolder);
       }
       ConfserverProcess.instance.emitter.once("valuesLoaded", resolve);
     });
@@ -220,7 +221,24 @@ export class ConfserverProcess {
     if (enableCCache) {
       reconfigureArgs.push("--ccache");
     }
-    reconfigureArgs.push("-C", currWorkspace.fsPath, "reconfigure");
+    reconfigureArgs.push("-C", currWorkspace.fsPath);
+    const sdkconfigDefaults =
+      (idfConf.readParameter("idf.sdkconfigDefaults") as string[]) || [];
+
+    if (reconfigureArgs.indexOf("SDKCONFIG") === -1) {
+      reconfigureArgs.push(`-DSDKCONFIG=${ConfserverProcess.instance.configFile}`)
+    }
+
+    if (
+      reconfigureArgs.indexOf("SDKCONFIG_DEFAULTS") === -1 &&
+      sdkconfigDefaults &&
+      sdkconfigDefaults.length
+    ) {
+      reconfigureArgs.push(
+        `-DSDKCONFIG_DEFAULTS='${sdkconfigDefaults.join(";")}'`
+      );
+    }
+    reconfigureArgs.push("reconfigure");
     const getSdkconfigProcess = spawn(pythonBinPath, reconfigureArgs, {
       env: modifiedEnv,
     });
@@ -299,6 +317,7 @@ export class ConfserverProcess {
     workspaceFolder: vscode.Uri,
     extensionPath: string,
     pythonBinPath: string,
+    configFile: string,
     modifiedEnv: { [key: string]: string }
   ) {
     this.workspaceFolder = workspaceFolder;
@@ -308,6 +327,7 @@ export class ConfserverProcess {
       idfConf.readParameter("idf.espIdfPath", workspaceFolder).toString() ||
       process.env.IDF_PATH;
     modifiedEnv.PYTHONUNBUFFERED = "0";
+    this.configFile = configFile;
     const idfPath = path.join(this.espIdfPath, "tools", "idf.py");
     const enableCCache = idfConf.readParameter(
       "idf.enableCCache",

--- a/src/espIdf/menuconfig/confServerProcess.ts
+++ b/src/espIdf/menuconfig/confServerProcess.ts
@@ -317,6 +317,27 @@ export class ConfserverProcess {
     if (enableCCache) {
       confServerArgs.push("--ccache");
     }
+    const buildDirPath = idfConf.readParameter(
+      "idf.buildPath",
+      workspaceFolder
+    ) as string;
+    confServerArgs.push("-B", buildDirPath);
+    const sdkconfigDefaults =
+      (idfConf.readParameter("idf.sdkconfigDefaults") as string[]) || [];
+
+    if (confServerArgs.indexOf("SDKCONFIG") === -1) {
+      confServerArgs.push(`-DSDKCONFIG=${this.configFile}`)
+    }
+
+    if (
+      confServerArgs.indexOf("SDKCONFIG_DEFAULTS") === -1 &&
+      sdkconfigDefaults &&
+      sdkconfigDefaults.length
+    ) {
+      confServerArgs.push(
+        `-DSDKCONFIG_DEFAULTS='${sdkconfigDefaults.join(";")}'`
+      );
+    }
     confServerArgs.push("-C", workspaceFolder.fsPath, "confserver");
     this.confServerProcess = spawn(pythonBinPath, confServerArgs, {
       env: modifiedEnv,

--- a/src/espIdf/reconfigure/task.ts
+++ b/src/espIdf/reconfigure/task.ts
@@ -27,7 +27,7 @@ import {
   workspace,
 } from "vscode";
 import { NotificationMode, readParameter } from "../../idfConfiguration";
-import { appendIdfAndToolsToPath } from "../../utils";
+import { appendIdfAndToolsToPath, compareVersion, getEspIdfFromCMake, getSDKConfigFilePath } from "../../utils";
 import { join } from "path";
 import { TaskManager } from "../../taskManager";
 import { getVirtualEnvPythonPath } from "../../pythonManager";
@@ -65,12 +65,23 @@ export class IdfReconfigureTask {
 
     const idfPy = join(this.idfPathDir, "tools", "idf.py");
     const reconfigureArgs = [idfPy];
+    const espIdfVersion = await getEspIdfFromCMake(this.idfPathDir);
+    const useEqualSign = compareVersion(espIdfVersion, "4.4") >= 0;
 
     let buildPathArgsIndex = reconfigureArgs.indexOf("-B");
     if (buildPathArgsIndex !== -1) {
-      reconfigureArgs.splice(buildPathArgsIndex, 2);
+      reconfigureArgs.splice(buildPathArgsIndex, useEqualSign ? 1 : 2);
     }
-    reconfigureArgs.push("-B", this.buildDirPath);
+    if (useEqualSign) {
+      reconfigureArgs.push(`-B=${this.buildDirPath}`);
+    } else {
+      reconfigureArgs.push("-B", this.buildDirPath);
+    }
+
+    const sdkconfigFile = await getSDKConfigFilePath(this.curWorkspace);
+    if (reconfigureArgs.indexOf("SDKCONFIG") === -1) {
+      reconfigureArgs.push(`-DSDKCONFIG=${sdkconfigFile}`);
+    }
 
     const sdkconfigDefaults =
       (readParameter("idf.sdkconfigDefaults") as string[]) || [];

--- a/src/espIdf/setTarget/setTargetInIdf.ts
+++ b/src/espIdf/setTarget/setTargetInIdf.ts
@@ -59,6 +59,16 @@ export async function setTargetInIDF(
   } else {
     modifiedEnv.IDF_CCACHE_ENABLE = undefined;
   }
+  if (modifiedEnv.SDKCONFIG) {
+    setTargetArgs.push(`-DSDKCONFIG='${modifiedEnv.SDKCONFIG}'`);
+  }
+  const sdkconfigDefaults =
+    (readParameter("idf.sdkconfigDefaults") as string[]) || [];
+
+  if (sdkconfigDefaults && sdkconfigDefaults.length) {
+    setTargetArgs.push(`-DSDKCONFIG_DEFAULTS='${sdkconfigDefaults.join(";")}'`);
+  }
+
   setTargetArgs.push("set-target", selectedTarget.target);
   const pythonBinPath = await getVirtualEnvPythonPath(workspaceFolder.uri);
   const setTargetResult = await spawn(pythonBinPath, setTargetArgs, {

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -143,7 +143,7 @@ import { getFileList, getTestComponents } from "./espIdf/unitTest/utils";
 import { saveDefSdkconfig } from "./espIdf/menuconfig/saveDefConfig";
 import { createSBOM, installEspSBOM } from "./espBom";
 import { getEspHomeKitSdk } from "./espHomekit/espHomekitDownload";
-import { selectIdfSetup } from "./versionSwitcher";
+import { getCurrentIdfSetup, selectIdfSetup } from "./versionSwitcher";
 import { checkDebugAdapterRequirements } from "./espIdf/debugAdapter/checkPyReqs";
 import { CDTDebugConfigurationProvider } from "./cdtDebugAdapter/debugConfProvider";
 import { CDTDebugAdapterDescriptorFactory } from "./cdtDebugAdapter/server";
@@ -383,6 +383,19 @@ export async function activate(context: vscode.ExtensionContext) {
             ESP.ProjectConfiguration.store.clear(
               ESP.ProjectConfiguration.SELECTED_CONFIG
             );
+          }
+          const currentIdfSetup = await getCurrentIdfSetup(
+            workspaceRoot,
+            false
+          );
+          if (statusBarItems["currentIdfVersion"]) {
+            statusBarItems["currentIdfVersion"].text = currentIdfSetup.isValid
+              ? `$(${
+                  commandDictionary[CommandKeys.SelectCurrentIdfVersion].iconId
+                }) ESP-IDF v${currentIdfSetup.version}`
+              : `$(${
+                  commandDictionary[CommandKeys.SelectCurrentIdfVersion].iconId
+                }) ESP-IDF InvalidSetup`;
           }
           const coverageOptions = getCoverageOptions(workspaceRoot);
           covRenderer = new CoverageRenderer(workspaceRoot, coverageOptions);
@@ -928,6 +941,16 @@ export async function activate(context: vscode.ExtensionContext) {
           ESP.ProjectConfiguration.store.clear(
             ESP.ProjectConfiguration.SELECTED_CONFIG
           );
+        }
+        const currentIdfSetup = await getCurrentIdfSetup(workspaceRoot, false);
+        if (statusBarItems["currentIdfVersion"]) {
+          statusBarItems["currentIdfVersion"].text = currentIdfSetup.isValid
+            ? `$(${
+                commandDictionary[CommandKeys.SelectCurrentIdfVersion].iconId
+              }) ESP-IDF v${currentIdfSetup.version}`
+            : `$(${
+                commandDictionary[CommandKeys.SelectCurrentIdfVersion].iconId
+              }) ESP-IDF InvalidSetup`;
         }
         const debugAdapterConfig = {
           currentWorkspace: workspaceRoot,
@@ -1817,6 +1840,7 @@ export async function activate(context: vscode.ExtensionContext) {
         workspaceRoot
       );
       await setIdfTarget(enterDeviceTargetMsg, workspaceFolder);
+      await getIdfTargetFromSdkconfig(workspaceRoot, statusBarItems["target"]);
     });
   });
 

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -373,6 +373,17 @@ export async function activate(context: vscode.ExtensionContext) {
               `$(${commandDictionary[CommandKeys.SelectSerialPort].iconId}) ` +
               idfConf.readParameter("idf.port", workspaceRoot);
           }
+          if (statusBarItems["projectConf"]) {
+            statusBarItems["projectConf"].dispose();
+            statusBarItems["projectConf"] = undefined;
+            const selectedConfig = ESP.ProjectConfiguration.store.get<string>(
+              ESP.ProjectConfiguration.SELECTED_CONFIG
+            );
+            ESP.ProjectConfiguration.store.clear(selectedConfig);
+            ESP.ProjectConfiguration.store.clear(
+              ESP.ProjectConfiguration.SELECTED_CONFIG
+            );
+          }
           const coverageOptions = getCoverageOptions(workspaceRoot);
           covRenderer = new CoverageRenderer(workspaceRoot, coverageOptions);
           break;
@@ -907,6 +918,17 @@ export async function activate(context: vscode.ExtensionContext) {
           tooltip: option.uri.fsPath,
         };
         utils.updateStatus(statusBarItems["workspace"], workspaceFolderInfo);
+        if (statusBarItems["projectConf"]) {
+          statusBarItems["projectConf"].dispose();
+          statusBarItems["projectConf"] = undefined;
+          const selectedConfig = ESP.ProjectConfiguration.store.get<string>(
+            ESP.ProjectConfiguration.SELECTED_CONFIG
+          );
+          ESP.ProjectConfiguration.store.clear(selectedConfig);
+          ESP.ProjectConfiguration.store.clear(
+            ESP.ProjectConfiguration.SELECTED_CONFIG
+          );
+        }
         const debugAdapterConfig = {
           currentWorkspace: workspaceRoot,
         } as IDebugAdapterConfig;

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -1090,6 +1090,7 @@ export async function activate(context: vscode.ExtensionContext) {
         commandDictionary[CommandKeys.SelectProjectConfiguration].checkboxState
       );
       await getIdfTargetFromSdkconfig(workspaceRoot, statusBarItems["target"]);
+      await utils.setCCppPropertiesJsonCompileCommands(workspaceRoot);
       ConfserverProcess.dispose();
     });
   });

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -1090,6 +1090,7 @@ export async function activate(context: vscode.ExtensionContext) {
         commandDictionary[CommandKeys.SelectProjectConfiguration].checkboxState
       );
       await getIdfTargetFromSdkconfig(workspaceRoot, statusBarItems["target"]);
+      ConfserverProcess.dispose();
     });
   });
 

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -71,7 +71,7 @@ import { ArduinoComponentInstaller } from "./espIdf/arduino/addArduinoComponent"
 import { PartitionTableEditorPanel } from "./espIdf/partition-table";
 import { ESPEFuseTreeDataProvider } from "./efuse/view";
 import { ESPEFuseManager } from "./efuse";
-import { constants, createFileSync, pathExists, readJson } from "fs-extra";
+import { constants, createFileSync, pathExists } from "fs-extra";
 import { getEspAdf } from "./espAdf/espAdfDownload";
 import { getEspMdf } from "./espMdf/espMdfDownload";
 import { SetupPanel } from "./setup/SetupPanel";

--- a/src/project-conf/index.ts
+++ b/src/project-conf/index.ts
@@ -43,13 +43,6 @@ export class ProjectConfigStore {
   public clear(key: string) {
     return this.set(key, undefined);
   }
-
-  public getKeys() {
-    return this.ctx.workspaceState.get<string[]>(
-      ESP.ProjectConfiguration.CONFIGURATION_LIST_KEY,
-      undefined
-    );
-  }
 }
 
 export async function getProjectConfigurationElements(workspaceFolder: Uri) {

--- a/src/project-conf/index.ts
+++ b/src/project-conf/index.ts
@@ -69,7 +69,6 @@ export async function getProjectConfigurationElements(workspaceFolder: Uri) {
       },
       env: projectConfJson[elem].env,
       flashBaudRate: projectConfJson[elem].flashBaudRate,
-      idfTarget: projectConfJson[elem].idfTarget,
       monitorBaudRate: projectConfJson[elem].monitorBaudRate,
       openOCD: {
         debugLevel: projectConfJson[elem].openOCD?.debugLevel,

--- a/src/project-conf/projectConfPanel.ts
+++ b/src/project-conf/projectConfPanel.ts
@@ -126,15 +126,6 @@ export class projectConfigurationPanel {
     this.panel.onDidDispose(() => this.dispose(), null, this.disposables);
   }
 
-  private async clearProjectConfFile() {
-    let projectConfKeys = ESP.ProjectConfiguration.store.getKeys();
-    if (projectConfKeys && projectConfKeys.length) {
-      for (const confKey of projectConfKeys) {
-        ESP.ProjectConfiguration.store.clear(confKey);
-      }
-    }
-  }
-
   private async openFolder() {
     const selectedFolder = await window.showOpenDialog({
       canSelectFolders: true,
@@ -163,7 +154,6 @@ export class projectConfigurationPanel {
   private async saveProjectConfFile(projectConfDict: {
     [key: string]: ProjectConfElement;
   }) {
-    this.clearProjectConfFile();
     const projectConfKeys = Object.keys(projectConfDict);
     this.clearSelectedProject(projectConfKeys);
     await saveProjectConfFile(this.workspaceFolder, projectConfDict);

--- a/src/project-conf/projectConfPanel.ts
+++ b/src/project-conf/projectConfPanel.ts
@@ -119,6 +119,17 @@ export class projectConfigurationPanel {
             });
           }
           break;
+        case "openFilePath":
+          let selectedFile = await this.openFile();
+          if (selectedFile) {
+            this.panel.webview.postMessage({
+              command: "setFilePath",
+              confKey: message.confKey,
+              newPath: selectedFile,
+              sectionsKeys: message.sectionsKeys,
+            });
+          }
+          break;
         default:
           break;
       }
@@ -134,6 +145,17 @@ export class projectConfigurationPanel {
     });
     if (selectedFolder && selectedFolder.length > 0) {
       return selectedFolder[0].fsPath;
+    }
+  }
+
+  private async openFile() {
+    const selectedFile = await window.showOpenDialog({
+      canSelectFolders: false,
+      canSelectFiles: true,
+      canSelectMany: false,
+    });
+    if (selectedFile && selectedFile.length > 0) {
+      return selectedFile[0].fsPath;
     }
   }
 

--- a/src/project-conf/projectConfiguration.ts
+++ b/src/project-conf/projectConfiguration.ts
@@ -26,7 +26,6 @@ export interface ProjectConfElement {
   };
   env: { [key: string]: string };
   flashBaudRate: string;
-  idfTarget: string;
   monitorBaudRate: string;
   openOCD: {
     debugLevel: number;

--- a/src/pythonManager.ts
+++ b/src/pythonManager.ts
@@ -269,6 +269,12 @@ export async function getVirtualEnvPythonPath(workspaceFolder: Uri) {
   let pythonPath = readParameter("idf.pythonInstallPath") as string;
   let espIdfDir = readParameter("idf.espIdfPath", workspaceFolder) as string;
   let idfToolsDir = readParameter("idf.toolsPath", workspaceFolder) as string;
+  const idfPathExists = await pathExists(espIdfDir);
+  const idfToolsPathExists = await pathExists(idfToolsDir);
+  const pythonPathExists = await pathExists(pythonPath);
+  if (!idfPathExists || !idfToolsPathExists || !pythonPathExists) {
+    return;
+  }
   const virtualEnvPython = await getPythonEnvPath(
     espIdfDir,
     idfToolsDir,

--- a/src/statusBar/index.ts
+++ b/src/statusBar/index.ts
@@ -29,6 +29,7 @@ import { readParameter } from "../idfConfiguration";
 import { ESP } from "../config";
 import { CommandItem } from "../cmdTreeView/cmdTreeDataProvider";
 import { CommandKeys, createCommandDictionary } from "../cmdTreeView/cmdStore";
+import { getIdfTargetFromSdkconfig } from "../workspaceConfig";
 
 export const statusBarItems: { [key: string]: StatusBarItem } = {};
 
@@ -54,14 +55,11 @@ export async function createCmdsStatusBarItems(workspaceFolder: Uri) {
     return {};
   }
   const port = readParameter("idf.port", workspaceFolder) as string;
-  let idfTarget = readParameter("idf.adapterTargetName", workspaceFolder);
+  let idfTarget = await getIdfTargetFromSdkconfig(workspaceFolder);
   let flashType = readParameter("idf.flashType", workspaceFolder) as string;
   let projectConf = ESP.ProjectConfiguration.store.get<string>(
     ESP.ProjectConfiguration.SELECTED_CONFIG
   );
-  if (idfTarget === "custom") {
-    idfTarget = readParameter("idf.customAdapterTargetName", workspaceFolder);
-  }
   let currentIdfVersion = await getCurrentIdfSetup(workspaceFolder, false);
 
 

--- a/src/support/writeReport.ts
+++ b/src/support/writeReport.ts
@@ -154,7 +154,6 @@ export async function writeTextReport(
         }
       }
       output += `Flash baud rate: ${reportedResult.projectConfigurations[key].flashBaudRate}${EOL}`;
-      output += `IDF Target: ${reportedResult.projectConfigurations[key].idfTarget}${EOL}`;
       output += `Monitor baud rate: ${reportedResult.projectConfigurations[key].monitorBaudRate}${EOL}`;
 
       if (reportedResult.projectConfigurations[key].openOCD) {

--- a/src/utils.ts
+++ b/src/utils.ts
@@ -241,15 +241,6 @@ export async function createVscodeFolder(curWorkspaceFsPath: vscode.Uri) {
 export async function setCCppPropertiesJsonCompilerPath(
   curWorkspaceFsPath: vscode.Uri
 ) {
-  const cCppPropertiesJsonPath = path.join(
-    curWorkspaceFsPath.fsPath,
-    ".vscode",
-    "c_cpp_properties.json"
-  );
-  const doesPathExists = await pathExists(cCppPropertiesJsonPath);
-  if (!doesPathExists) {
-    return;
-  }
   const modifiedEnv = await appendIdfAndToolsToPath(curWorkspaceFsPath);
   const idfTarget = modifiedEnv.IDF_TARGET || "esp32";
   const gccTool = getToolchainToolName(idfTarget, "gcc");
@@ -258,22 +249,65 @@ export async function setCCppPropertiesJsonCompilerPath(
     curWorkspaceFsPath.fsPath,
     modifiedEnv
   );
+  if (!compilerAbsolutePath) {
+    return;
+  }
+  let compilerRelativePath = compilerAbsolutePath.split(
+    modifiedEnv.IDF_TOOLS_PATH
+  )[1];
+  const settingToUse =
+    process.platform === "win32"
+      ? "${config:idf.toolsPathWin}"
+      : "${config:idf.toolsPath}";
 
+  await updateCCppPropertiesJson(
+    curWorkspaceFsPath,
+    "compilerPath",
+    settingToUse + compilerRelativePath
+  );
+}
+
+export async function setCCppPropertiesJsonCompileCommands(
+  curWorkspaceFsPath: vscode.Uri
+) {
+  const buildDirPath = idfConf.readParameter(
+    "idf.buildPath",
+    curWorkspaceFsPath
+  ) as string;
+  const compileCommandsPath = path.join(buildDirPath, "compile_commands.json");
+
+  await updateCCppPropertiesJson(
+    curWorkspaceFsPath,
+    "compileCommands",
+    compileCommandsPath
+  );
+}
+
+export async function updateCCppPropertiesJson(
+  workspaceUri: vscode.Uri,
+  fieldToUpdate: string,
+  newFieldValue: string
+) {
+  const cCppPropertiesJsonPath = path.join(
+    workspaceUri.fsPath,
+    ".vscode",
+    "c_cpp_properties.json"
+  );
+  const doesPathExists = await pathExists(cCppPropertiesJsonPath);
+  if (!doesPathExists) {
+    return;
+  }
   const cCppPropertiesJson = await readJSON(cCppPropertiesJsonPath);
   if (
     cCppPropertiesJson &&
     cCppPropertiesJson.configurations &&
     cCppPropertiesJson.configurations.length
   ) {
-    let compilerRelativePath = compilerAbsolutePath.split(
-      modifiedEnv.IDF_TOOLS_PATH
-    )[1];
-    const settingToUse =
-      process.platform === "win32"
-        ? "${config:idf.toolsPathWin}"
-        : "${config:idf.toolsPath}";
-    cCppPropertiesJson.configurations[0].compilerPath =
-      settingToUse + compilerRelativePath;
+    const buildDirPath = idfConf.readParameter(
+      "idf.buildPath",
+      workspaceUri
+    ) as string;
+    cCppPropertiesJson.configurations[0][fieldToUpdate] = newFieldValue;
     await writeJSON(cCppPropertiesJsonPath, cCppPropertiesJson, {
       spaces: vscode.workspace.getConfiguration().get("editor.tabSize") || 2,
     });

--- a/src/utils.ts
+++ b/src/utils.ts
@@ -1109,7 +1109,8 @@ export async function appendIdfAndToolsToPath(curWorkspace: vscode.Uri) {
   }
 
   let sdkconfigFilePath = idfConf.readParameter(
-    "idf.sdkconfigFilePath"
+    "idf.sdkconfigFilePath",
+    curWorkspace
   ) as string;
   if (sdkconfigFilePath) {
     modifiedEnv.SDKCONFIG = sdkconfigFilePath;

--- a/src/views/project-conf/components/projectConfElem.vue
+++ b/src/views/project-conf/components/projectConfElem.vue
@@ -107,13 +107,6 @@ function removeValueFromArray(sections: string[], index: number) {
       :sections="['flashBaudRate']"
       :updateMethod="updateElement"
     />
-    <SelectElement
-      :selectValue="el.idfTarget ? el.idfTarget : ''"
-      title="IDF Target"
-      :options="idfTargets"
-      :sections="['idfTarget']"
-      :updateMethod="updateElement"
-    />
     <StringElement
       title="Monitor baud rate"
       :value="el.monitorBaudRate ? el.monitorBaudRate : ''"

--- a/src/views/project-conf/components/projectConfElem.vue
+++ b/src/views/project-conf/components/projectConfElem.vue
@@ -21,22 +21,16 @@ const openOcdDebugLevelOptions: { name: string; value: number }[] = [
   { name: "Verbose", value: 4 },
 ];
 
-const idfTargets: { name: string; value: string }[] = [
-  { name: "esp32", value: "esp32" },
-  { name: "ESP32 S2", value: "esp32s2" },
-  { name: "ESP32 S3", value: "esp32s3" },
-  { name: "ESP32 C2", value: "esp32c2" },
-  { name: "ESP32 C3", value: "esp32c3" },
-  { name: "ESP32 C6", value: "esp32c6" },
-  { name: "ESP32 H2", value: "esp32h2" },
-];
-
 function updateElement(sections: string[], newValue: any) {
   store.updateConfigElement({ confKey: props.title, sections, newValue });
 }
 
 function openBuildDir(sections: string[]) {
   store.openBuildPath({ confKey: props.title, sections });
+}
+
+function openFilePath(sections: string[]) {
+  store.openFilePath({ confKey: props.title, sections });
 }
 
 function addValueToArray(sections: string[], newValue: any) {
@@ -93,6 +87,7 @@ function removeValueFromArray(sections: string[], index: number) {
         "
         :sections="['build', 'sdkconfigFilePath']"
         :updateMethod="updateElement"
+        :openMethod="openFilePath"
       />
     </div>
     <DictionaryElement

--- a/src/views/project-conf/main.ts
+++ b/src/views/project-conf/main.ts
@@ -2,13 +2,13 @@
  * Project: ESP-IDF VSCode Extension
  * File Created: Thursday, 31st August 2023 12:31:23 pm
  * Copyright 2023 Espressif Systems (Shanghai) CO LTD
- * 
+ *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
- * 
+ *
  *    http://www.apache.org/licenses/LICENSE-2.0
- * 
+ *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
@@ -38,6 +38,15 @@ window.addEventListener("message", (event: MessageEvent) => {
           confKey: msg.confKey,
           sections: msg.sectionsKeys,
           newValue: msg.buildPath,
+        });
+      }
+      break;
+    case "setFilePath":
+      if (msg.confKey) {
+        store.updateConfigElement({
+          confKey: msg.confKey,
+          sections: msg.sectionsKeys,
+          newValue: msg.newPath,
         });
       }
       break;

--- a/src/views/project-conf/store.ts
+++ b/src/views/project-conf/store.ts
@@ -86,6 +86,14 @@ export const useProjectConfStore = defineStore("project-config", () => {
     });
   }
 
+  function openFilePath(payload: { confKey: string; sections: string[] }) {
+    vscode.postMessage({
+      command: "openFilePath",
+      sectionsKeys: payload.sections,
+      confKey: payload.confKey,
+    });
+  }
+
   function addNewConfigToList(confKey: string) {
     let newConf = {
       build: {
@@ -174,6 +182,7 @@ export const useProjectConfStore = defineStore("project-config", () => {
     addNewConfigToList,
     addValueToConfigElement,
     openBuildPath,
+    openFilePath,
     updateConfigElement,
     removeValueFromConfigElement,
     requestInitValues,

--- a/src/workspaceConfig.ts
+++ b/src/workspaceConfig.ts
@@ -82,21 +82,23 @@ export async function getIdfTargetFromSdkconfig(
   workspacePath: vscode.Uri,
   statusItem?: vscode.StatusBarItem
 ) {
-  let sdkConfigPath = await getSDKConfigFilePath(workspacePath);
-  const doesSdkconfigExists = await pathExists(sdkConfigPath);
-  if (!doesSdkconfigExists) {
-    return;
+  try {
+    const configIdfTarget = await utils.getConfigValueFromSDKConfig(
+      "CONFIG_IDF_TARGET",
+      workspacePath
+    );
+    let idfTarget = configIdfTarget.replace(/\"/g, "");
+    if (!idfTarget) {
+      idfTarget = "esp32";
+    }
+    if (statusItem) {
+      statusItem.text = "$(chip) " + idfTarget;
+    }
+    return idfTarget;
+  } catch (error) {
+    if (statusItem) {
+      statusItem.text = "$(chip) esp32";
+    }
+    return "esp32";
   }
-  const configIdfTarget = await utils.getConfigValueFromSDKConfig(
-    "CONFIG_IDF_TARGET",
-    workspacePath
-  );
-  const idfTarget = configIdfTarget.replace(/\"/g, "");
-  if (!idfTarget) {
-    return;
-  }
-  if (statusItem) {
-    statusItem.text = "$(chip) " + idfTarget;
-  }
-  return idfTarget;
 }

